{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Analysis of DNA using `BioPython`\n",
    "\n",
    "Building blocks of DNA and RNA:\n",
    "\n",
    "* `A` Adenine\n",
    "* `C` Cytosine\n",
    "* `G` Guanine\n",
    "* `T` Thymine  \n",
    "  \n",
    "\n",
    "* `U` Uracil (in RNA Thymine → Uracil)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a General DNA sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA')"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import Bio\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "mydna = Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA')\n",
    "mydna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a Specific Sequence (DNA,RNA,Protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RNAAlphabet()"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Alphabet import generic_dna,generic_rna,generic_protein\n",
    "\n",
    "# Create a RNA\n",
    "rna1 = Seq('AGGCUCUCGUA',generic_rna)\n",
    "rna1.alphabet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequence Manipulation\n",
    "* Indexing/Slicing\n",
    "* Join 2 Sequences\n",
    "* Find a Codon in a sequence\n",
    "* Count the number of Nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATT', DNAAlphabet())"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq = Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA',generic_dna)\n",
    "dna_seq[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of G Nucleotides in a sequence\n",
    "dna_seq.count('GGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the index/position of G Nucleotides in a sequence\n",
    "dna_seq.find('G')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of G Nucleotides in a sequence in that overlap\n",
    "dna_seq.count_overlap('GGT')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequence Compliments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATCTCGTAA')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq1 = Seq('ATGATCTCGTAA')\n",
    "seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TACTAGAGCATT')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq1.complement()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Transcription to `RNA`\n",
    "Same base pairs as DNA, but Thymine changes to Uracil `T → U`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq('ATGATCTCGTAA')\n",
      "Seq('AUGAUCUCGUAA', RNAAlphabet())\n"
     ]
    }
   ],
   "source": [
    "# To mrna\n",
    "print(seq1.__repr__())\n",
    "print(seq1.transcribe().__repr__())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AUGAUCUCGUAA', RNAAlphabet())"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transcription to RNA\n",
    "seq1.transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq1.transcribe().translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Translate to `Protein`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq('ATGATCTCGTAA')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Seq('MIS*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(seq1.__repr__())\n",
    "\n",
    "# Translate RNA to protein\n",
    "seq1.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', DNAAlphabet())\n",
      "Seq('AUUAAAGGUUUAUACCUUCCCAGGUAACAAACCAACCAACUUUCGAUCUCUUGU...AAA', RNAAlphabet())\n"
     ]
    }
   ],
   "source": [
    "mrna = dna_seq.transcribe()\n",
    "\n",
    "print(repr(dna_seq))\n",
    "print(repr(mrna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/anaconda3/envs/dna-toolbox/lib/python3.7/site-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Seq('IKGLYLPR*QTNQLSISCRSVL*', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate mRNA to Protein/Amino Acid\n",
    "mrna.translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stop codons between Proteins\n",
    "There are 3 STOP codons in the genetic code\n",
    "`UAG`, `UAA`, and `UGA`.  \n",
    "These codons signal the end of the polypeptide chain during translation\n",
    "\n",
    "> Polypeptides are chains of amino acids. Proteins are made up of one or more polypeptide molecules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('IKGLYLPR@QTNQLSISCRSVL@', HasStopCodon(ExtendedIUPACProtein(), '@'))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate mRNA to Protein/Amino Acid\n",
    "# Change the symbol for the stop codon\n",
    "mrna.translate(stop_symbol='@')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', DNAAlphabet())"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Back Transcribe mRNA to DNA\n",
    "mrna.back_transcribe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Translate DAN Sequence to `Condon`\n",
    "> A codon is a sequence of three DNA or RNA nucleotides that corresponds with a specific amino acid or stop signal during protein synthesis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Custom Translate\n",
    "def custom_translate(seq): \n",
    "      \n",
    "    table = { \n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_', \n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W', \n",
    "    } \n",
    "    protein =\"\" \n",
    "    if len(seq)%3 == 0: \n",
    "        for i in range(0, len(seq), 3): \n",
    "            codon = seq[i:i + 3] \n",
    "            protein+= table[codon] \n",
    "    return protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', DNAAlphabet())"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a DNA\n",
    "dna1 = Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA',generic_dna)\n",
    "dna1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATTAAAGGTTTATAC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/anaconda3/envs/dna-toolbox/lib/python3.7/site-packages/Bio/Seq.py:182: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  BiopythonWarning,\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'IKGLY'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(dna1[0:15])\n",
    "custom_translate(dna1[0:15])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "from Bio.Data import CodonTable\n",
    "\n",
    "# CodonTable for DNA\n",
    "print(CodonTable.unambiguous_dna_by_name['Standard'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  U      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "U | UUU F   | UCU S   | UAU Y   | UGU C   | U\n",
      "U | UUC F   | UCC S   | UAC Y   | UGC C   | C\n",
      "U | UUA L   | UCA S   | UAA Stop| UGA Stop| A\n",
      "U | UUG L(s)| UCG S   | UAG Stop| UGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CUU L   | CCU P   | CAU H   | CGU R   | U\n",
      "C | CUC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CUA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CUG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | AUU I   | ACU T   | AAU N   | AGU S   | U\n",
      "A | AUC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | AUA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | AUG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GUU V   | GCU A   | GAU D   | GGU G   | U\n",
      "G | GUC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GUA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GUG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "# CodonTable for RNA\n",
    "print(CodonTable.unambiguous_rna_by_name['Standard'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Covid19 (`SARS-CoV-2`)\n",
    "\n",
    "GenBank provides a list the SARS-CoV-2 sequences currently available from the Sequence Read Archive (SRA). At the time of writing this there are `2124` Nucleotide Sequences for SARS-CoV-2.\n",
    "\n",
    "* https://www.ncbi.nlm.nih.gov/genbank/sars-cov-2-seqs/\n",
    "\n",
    "The sample in this example is `NC_045512` from China collected in December 2019.\n",
    "\n",
    "> Nature: `A new coronavirus associated with human respiratory disease in China`  \n",
    "> Received: 7 January 2020  \n",
    "> https://doi.org/10.1038/s41586-020-2008-3\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: MN908947.3\n",
      "Name: MN908947.3\n",
      "Description: MN908947.3 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Number of features: 0\n",
      "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "for record in SeqIO.parse(\"../data/MN908947.fna\", \"fasta\"):\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the sequence record in the file\n",
    "ncov_dna_record = SeqIO.read(\"../data/MN908947.fna\",\"fasta\")\n",
    "ncov_dna = ncov_dna_record.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. `SARS-CoV-2` DNA Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA', SingleLetterAlphabet())"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the Nucleotides\n",
    "ncov_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29903"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Length of our sequence\n",
    "len(ncov_dna)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Transcription of DNA to RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AUUAAAGGUUUAUACCUUCCCAGGUAACAAACCAACCAACUUUCGAUCUCUUGU...AAA', RNAAlphabet())"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transcribe (DNA to mRNA)\n",
    "ncov_mRNA = ncov_dna.transcribe()\n",
    "\n",
    "# Changes Thymine to Uracil\n",
    "ncov_mRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. Translation from RNA to Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/anaconda3/envs/dna-toolbox/lib/python3.7/site-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Seq('IKGLYLPR*QTNQLSISCRSVL*TNFKICVAVTRLHA*CTHAV*LITNYCR*QD...KKK', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate to Protein/Amino Acids (mRNA to AA)\n",
    "ncov_protein = ncov_mRNA.translate()\n",
    "ncov_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9967"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Length of Protein/Amino Acids\n",
    "len(ncov_protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Over the sequence of DNA, we translated each set of _three nucleotides_ (a `codon`), into it's amino acid (protein)\n",
    "> A codon is a sequence of three DNA or RNA nucleotides that corresponds with a specific amino acid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9967.666666666666\n",
      "9967\n"
     ]
    }
   ],
   "source": [
    "# Check if it is true by dividing total length of neucleotides by the number of codons\n",
    "print(len(ncov_mRNA)/3)\n",
    "print(len(ncov_protein))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. Find Proteins (Amino Acid chains)\n",
    "\n",
    "Amino acids are the monomers that make up proteins. Specifically, a protein is made up of one or more linear chains of amino acids, each of which is called a `polypeptide`.\n",
    "\n",
    "> Stop codons (`UAG`, `UAA`, and `UGA`) between proteins signal the end of the polypeptide chain during translation.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total proteins: 775\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[Seq('IKGLYLPR', HasStopCodon(ExtendedIUPACProtein(), '*')),\n",
       " Seq('QTNQLSISCRSVL', HasStopCodon(ExtendedIUPACProtein(), '*')),\n",
       " Seq('TNFKICVAVTRLHA', HasStopCodon(ExtendedIUPACProtein(), '*')),\n",
       " Seq('CTHAV', HasStopCodon(ExtendedIUPACProtein(), '*')),\n",
       " Seq('LITNYCR', HasStopCodon(ExtendedIUPACProtein(), '*'))]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find all the amino acids\n",
    "ncov_amino_acids = ncov_protein.split('*')\n",
    "print('Total proteins:', len(ncov_amino_acids))\n",
    "\n",
    "# Print first 5\n",
    "ncov_amino_acids[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create DataFrame of Polypeptide Chains (Proteins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(775, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IKGLYLPR</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QTNQLSISCRSVL</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TNFKICVAVTRLHA</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTHAV</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LITNYCR</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      amino_acids  length\n",
       "0        IKGLYLPR       8\n",
       "1   QTNQLSISCRSVL      13\n",
       "2  TNFKICVAVTRLHA      14\n",
       "3           CTHAV       5\n",
       "4         LITNYCR       7"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Place our Amino Acids into a DataFrame\n",
    "df = pd.DataFrame({'amino_acids': [str(i) for i in ncov_amino_acids]})\n",
    "df['length'] = df['amino_acids'].apply(len)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>CTIVFKRVCGVSAARLTPCGTGTSTDVVYRAFDIYNDKVAGFAKFL...</td>\n",
       "      <td>2701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>ASAQRSQITLHINELMDLFMRIFTIGTVTLKQGEIKDATPSDFVRA...</td>\n",
       "      <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>TNMKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNS...</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>AQADEYELMYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALR...</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>QQMFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSL...</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DGEPCPWFQRENTRPTQFACFTGSRRARTWLWRLRGGGLIRGTSTS</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>TMLRCYFPKCSEKNNQGYTPLVVTHNFDFTFSFSPEYSMVFVLFFV</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>DVVYTHWYWSGNNSYTGSQYGSRILWWCIVLSVLPLPHRSSKS</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>LQTLAANCTICPQRFSVLRNVAHWHGSHTFGNVVDLHRCHQIG</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771</th>\n",
       "      <td>KSHHIFTEATRSTIECTVNNARESCLYGRALMCKINFSSAIPM</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           amino_acids  length\n",
       "548  CTIVFKRVCGVSAARLTPCGTGTSTDVVYRAFDIYNDKVAGFAKFL...    2701\n",
       "694  ASAQRSQITLHINELMDLFMRIFTIGTVTLKQGEIKDATPSDFVRA...     290\n",
       "719  TNMKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNS...     123\n",
       "695  AQADEYELMYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALR...      83\n",
       "718  QQMFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSL...      63\n",
       "6       DGEPCPWFQRENTRPTQFACFTGSRRARTWLWRLRGGGLIRGTSTS      46\n",
       "464     TMLRCYFPKCSEKNNQGYTPLVVTHNFDFTFSFSPEYSMVFVLFFV      46\n",
       "539        DVVYTHWYWSGNNSYTGSQYGSRILWWCIVLSVLPLPHRSSKS      43\n",
       "758        LQTLAANCTICPQRFSVLRNVAHWHGSHTFGNVVDLHRCHQIG      43\n",
       "771        KSHHIFTEATRSTIECTVNNARESCLYGRALMCKINFSSAIPM      43"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Longest polypeptide chains\n",
    "df.nlargest(10, 'length')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f206cde0890>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAD4CAYAAAD1jb0+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAM2klEQVR4nO3df6id913A8ffHZkOXK2lq5iWm1VuxVELDWnuZlYmcu06J61grjLkyR6Yd1z9WrVKRuH9UZBBBNv1DxLDW5Q/tddSOlkY2Sta7KkjxZh0mbSwdXbo1pMtK29jbP5zRj3+cJ+56e26eJ+ecm3M+zfsF5Z7nOT/ulw9P3j15ch5OZCaSpHp+YNILkCQNx4BLUlEGXJKKMuCSVJQBl6SitlzKX7Zjx46cm5vr/Pg33niDrVu3bt6C3gKcUTtn1I1zajepGR09evTlzHzn+v2XNOBzc3OsrKx0fvzy8jK9Xm/zFvQW4IzaOaNunFO7Sc0oIl4YtN9TKJJUlAGXpKIMuCQVZcAlqSgDLklFGXBJKsqAS1JRBlySijLgklTUJb0ScxRz+w8P/dyTB24b40okaTr4DlySijLgklSUAZekogy4JBVlwCWpKAMuSUUZcEkqyoBLUlEGXJKKMuCSVJQBl6SiDLgkFWXAJakoAy5JRbUGPCKuiYjHI+KZiHg6Iu5p9l8VEY9FxHPNz+2bv1xJ0nld3oGfA+7NzN3ALcAnI2I3sB84kpnXAUeabUnSJdIa8Mw8nZlfa26/DpwAdgG3A4eahx0C7tisRUqS3uyizoFHxBxwE/AkMJuZp5u7XgJmx7oySdIFRWZ2e2DEDPBV4NOZ+VBEvJaZV665/9XMfNN58IhYBBYBZmdnb15aWuq8uNXVVWZmZgA4dups5+ett2fXtqGfO+3WzkiDOaNunFO7Sc1oYWHhaGbOr9/fKeAR8TbgUeDLmfmZZt+zQC8zT0fETmA5M6+/0OvMz8/nyspK50UvLy/T6/UAvxNzI2tnpMGcUTfOqd2kZhQRAwPe5VMoAdwHnDgf78YjwL7m9j7g4XEsVJLUTZdvpX8P8DHgWER8vdn3KeAA8IWIuAt4Afjw5ixRkjRIa8Az85+B2ODuW8e7HElSV16JKUlFGXBJKsqAS1JRBlySijLgklSUAZekogy4JBVlwCWpKAMuSUUZcEkqyoBLUlEGXJKKMuCSVJQBl6SiDLgkFWXAJakoAy5JRRlwSSrKgEtSUQZckooy4JJUlAGXpKIMuCQVZcAlqSgDLklFGXBJKsqAS1JRBlySijLgklSUAZekogy4JBVlwCWpKAMuSUUZcEkqyoBLUlEGXJKKMuCSVJQBl6SiWgMeEfdHxJmIOL5m3x9FxKmI+Hrz3/s3d5mSpPW6vAP/PLB3wP7PZuaNzX//ON5lSZLatAY8M58AXrkEa5EkXYRRzoHfHRH/1pxi2T62FUmSOonMbH9QxBzwaGbe0GzPAi8DCfwJsDMzf2OD5y4CiwCzs7M3Ly0tdV7c6uoqMzMzABw7dbbz89bbs2vb0M+ddmtnpMGcUTfOqd2kZrSwsHA0M+fX7x8q4F3vW29+fj5XVlY6LLdveXmZXq8HwNz+w52ft97JA7cN/dxpt3ZGGswZdeOc2k1qRhExMOBDnUKJiJ1rNn8FOL7RYyVJm2NL2wMi4gGgB+yIiBeBPwR6EXEj/VMoJ4Hf3MQ1SpIGaA14Zt45YPd9m7AWSdJF8EpMSSrKgEtSUQZckooy4JJUlAGXpKIMuCQVZcAlqSgDLklFGXBJKsqAS1JRBlySijLgklSUAZekogy4JBVlwCWpKAMuSUUZcEkqyoBLUlEGXJKKMuCSVJQBl6SiDLgkFWXAJakoAy5JRRlwSSrKgEtSUQZckooy4JJUlAGXpKIMuCQVZcAlqSgDLklFGXBJKsqAS1JRBlySijLgklSUAZekogy4JBVlwCWpqNaAR8T9EXEmIo6v2XdVRDwWEc81P7dv7jIlSet1eQf+eWDvun37gSOZeR1wpNmWJF1CrQHPzCeAV9btvh041Nw+BNwx5nVJklpEZrY/KGIOeDQzb2i2X8vMK5vbAbx6fnvAcxeBRYDZ2dmbl5aWOi9udXWVmZkZAI6dOtv5eevt2bVt6OdOu7Uz0mDOqBvn1G5SM1pYWDiamfPr928Z9YUzMyNiw/8LZOZB4CDA/Px89nq9zq+9vLzM+cd/fP/hodd48qPdf2c1a2ekwZxRN86p3bTNaNhPoXwnInYCND/PjG9JkqQuhg34I8C+5vY+4OHxLEeS1FWXjxE+APwLcH1EvBgRdwEHgF+MiOeA9zXbkqRLqPUceGbeucFdt455LZKki+CVmJJUlAGXpKIMuCQVZcAlqSgDLklFGXBJKsqAS1JRBlySijLgklSUAZekogy4JBVlwCWpqJG/0KGCuRG+DALg5IHbxrQSSRof34FLUlEGXJKKMuCSVJQBl6SiDLgkFWXAJakoAy5JRRlwSSrKgEtSUQZckooy4JJUlAGXpKIMuCQVZcAlqSgDLklFGXBJKsqAS1JRBlySijLgklSUAZekogy4JBVlwCWpKAMuSUUZcEkqyoBLUlFbRnlyRJwEXgf+GziXmfPjWJQkqd1IAW8sZObLY3gdSdJF8BSKJBUVmTn8kyO+CbwKJPDXmXlwwGMWgUWA2dnZm5eWljq//urqKjMzMwAcO3V26HWOas+ubRP73W3WzkiDOaNunFO7Sc1oYWHh6KBT1KMGfFdmnoqIHwUeA34rM5/Y6PHz8/O5srLS+fWXl5fp9XoAzO0/PPQ6R3XywG0T+91t1s5IgzmjbpxTu0nNKCIGBnykUyiZear5eQb4IvDuUV5PktTd0AGPiK0R8cPnbwO/BBwf18IkSRc2yqdQZoEvRsT51/m7zPzSWFYlSWo1dMAz83ngXWNciyTpIvgxQkkqyoBLUlEGXJKKMuCSVJQBl6SiDLgkFWXAJakoAy5JRRlwSSrKgEtSUQZckooy4JJUlAGXpKLG8aXGb3mjfBvQNH+bj6Tu5vYf5t495/j4kD3YjBb4DlySijLgklSUAZekogy4JBVlwCWpKAMuSUUZcEkqyoBLUlFeyDPFulxAtNGFBaNcNDDKhUuj/m5J3fkOXJKKMuCSVJQBl6SiDLgkFWXAJakoAy5JRRlwSSrKgEtSUV7Is8lGvSimokl9g9FGv3eUb1GZdlUvmvJbrsbDd+CSVJQBl6SiDLgkFWXAJakoAy5JRRlwSSpqpIBHxN6IeDYivhER+8e1KElSu6EDHhFXAH8J/DKwG7gzInaPa2GSpAsb5R34u4FvZObzmfk9YAm4fTzLkiS1icwc7okRHwL2ZuYnmu2PAT+bmXeve9wisNhsXg88exG/Zgfw8lALvHw4o3bOqBvn1G5SM/qJzHzn+p2bfil9Zh4EDg7z3IhYycz5MS/pLcUZtXNG3TindtM2o1FOoZwCrlmzfXWzT5J0CYwS8H8FrouIayPi7cBHgEfGsyxJUpuhT6Fk5rmIuBv4MnAFcH9mPj22lfUNderlMuOM2jmjbpxTu6ma0dD/iClJmiyvxJSkogy4JBU1lQH3Ev03i4hrIuLxiHgmIp6OiHua/VdFxGMR8Vzzc/uk1zppEXFFRDwVEY8229dGxJPN8fT3zT+6X9Yi4sqIeDAi/j0iTkTEz3ks/X8R8bvNn7XjEfFARPzgtB1LUxdwL9Hf0Dng3szcDdwCfLKZy37gSGZeBxxpti939wAn1mz/KfDZzPwp4FXgromsarr8BfClzPxp4F305+Wx1IiIXcBvA/OZeQP9D2p8hCk7lqYu4HiJ/kCZeTozv9bcfp3+H7hd9GdzqHnYIeCOyaxwOkTE1cBtwOea7QDeCzzYPMQZRWwDfgG4DyAzv5eZr+GxtN4W4IciYgvwDuA0U3YsTWPAdwHfXrP9YrNPjYiYA24CngRmM/N0c9dLwOyEljUt/hz4feB/mu0fAV7LzHPNtscTXAt8F/ib5lTT5yJiKx5L/yczTwF/BnyLfrjPAkeZsmNpGgOuC4iIGeAfgN/JzP9Ye1/2PxN62X4uNCI+AJzJzKOTXsuU2wL8DPBXmXkT8AbrTpd4LMV2+n8juRb4MWArsHeiixpgGgPuJfobiIi30Y/332bmQ83u70TEzub+ncCZSa1vCrwH+GBEnKR/6u299M/1Xtn8NRg8nqD/zvHFzHyy2X6QftA9lr7vfcA3M/O7mflfwEP0j6+pOpamMeBeoj9Acy73PuBEZn5mzV2PAPua2/uAhy/12qZFZv5BZl6dmXP0j5uvZOZHgceBDzUPu6xnBJCZLwHfjojrm123As/gsbTWt4BbIuIdzZ+98zOaqmNpKq/EjIj30z+Xef4S/U9PeEkTFxE/D/wTcIzvn9/9FP3z4F8Afhx4AfhwZr4ykUVOkYjoAb+XmR+IiJ+k/478KuAp4Ncy8z8nub5Ji4gb6f9D79uB54Ffp/+GzmOpERF/DPwq/U+APQV8gv4576k5lqYy4JKkdtN4CkWS1IEBl6SiDLgkFWXAJakoAy5JRRlwSSrKgEtSUf8LlX+JCexean0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.length.value_counts().hist(bins=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### What is the Most Abundent Amino Acid?\n",
    "> Because amino acids can be arranged in many different combinations, it's possible for your body to make thousands of different kinds of proteins from just the same 20/21/22 amino acids.\n",
    "\n",
    "* Throughout known life, there are 22 genetically encoded (proteinogenic) amino acids, 20 in the standard genetic code and an additional 2 that can be incorporated by special translation mechanisms.   \n",
    "* In eukaryotes, there are only 21 proteinogenic amino acids, the 20 of the standard genetic code, plus selenocysteine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Different amino acids: 21\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "L    886\n",
       "S    810\n",
       "*    774\n",
       "T    679\n",
       "C    635\n",
       "F    593\n",
       "R    558\n",
       "V    548\n",
       "Y    505\n",
       "N    472\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aa = pd.Series([c for c in str(ncov_protein)]).value_counts()\n",
    "print('Different amino acids:', aa.shape[0])\n",
    "display(aa.head(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3D Structure of `SARS-CoV-2`\n",
    "From `Protein Data Bank` https://www.ncbi.nlm.nih.gov/Structure/pdb/6LU7\n",
    "\n",
    "Requirements:\n",
    "```\n",
    "pip install nglview\n",
    "pip install py3Dmol\n",
    "\n",
    "jupyter-nbextension enable nglview --py --sys-prefix\n",
    "\n",
    "nglview enable\n",
    "jupyter-labextension install @jupyter-widget/jupyterlab-manager\n",
    "jupyter-labextension install nglview-js-widgets\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB import PDBParser, MMCIFParser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/anaconda3/envs/dna-toolbox/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2842.\n",
      "  PDBConstructionWarning,\n"
     ]
    }
   ],
   "source": [
    "# Reading a PDB File\n",
    "parser = PDBParser()\n",
    "covid_structure = parser.get_structure(\"mmdb_6LU7\", \"../data/6lu7.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Structure id=mmdb_6LU7>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covid_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "70c13b7edb1a4cabb38092cc36862e60",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# View 3D Structure with nglview\n",
    "import nglview as nv\n",
    "view =  nv.show_biopython(covid_structure)\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_15888699439063692\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_15888699439063692\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.csb.pitt.edu/build/3Dmol.js');\n}\n\nvar viewer_15888699439063692 = null;\nvar warn = document.getElementById(\"3dmolwarning_15888699439063692\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_15888699439063692 = $3Dmol.createViewer($(\"#3dmolviewer_15888699439063692\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:6LU7\", viewer_15888699439063692, {}, function() {\n\tviewer_15888699439063692.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\nviewer_15888699439063692.render();\n})\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_15888699439063692\"  style=\"position: relative; width: 640px; height: 480px\">\n",
       "        <p id=\"3dmolwarning_15888699439063692\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "      resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://3dmol.csb.pitt.edu/build/3Dmol.js');\n",
       "}\n",
       "\n",
       "var viewer_15888699439063692 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_15888699439063692\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_15888699439063692 = $3Dmol.createViewer($(\"#3dmolviewer_15888699439063692\"),{backgroundColor:\"white\"});\n",
       "$3Dmol.download(\"pdb:6LU7\", viewer_15888699439063692, {}, function() {\n",
       "\tviewer_15888699439063692.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\n",
       "viewer_15888699439063692.render();\n",
       "})\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x7f20a8510d10>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# View 3D Structure with py3Dmol\n",
    "import py3Dmol\n",
    "view1 = py3Dmol.view(query='pdb:6LU7')\n",
    "view1.setStyle({'cartoon':{'color':'spectrum'}})\n",
    "view1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_15888699724689698\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_15888699724689698\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.csb.pitt.edu/build/3Dmol.js');\n}\n\nvar viewer_15888699724689698 = null;\nvar warn = document.getElementById(\"3dmolwarning_15888699724689698\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_15888699724689698 = $3Dmol.createViewer($(\"#3dmolviewer_15888699724689698\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:6LU7\", viewer_15888699724689698, {}, function() {\n\tviewer_15888699724689698.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\n\tviewer_15888699724689698.render_image();\nviewer_15888699724689698.render();\n})\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_15888699724689698\"  style=\"position: relative; width: 640px; height: 480px\">\n",
       "        <p id=\"3dmolwarning_15888699724689698\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "      resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://3dmol.csb.pitt.edu/build/3Dmol.js');\n",
       "}\n",
       "\n",
       "var viewer_15888699724689698 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_15888699724689698\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_15888699724689698 = $3Dmol.createViewer($(\"#3dmolviewer_15888699724689698\"),{backgroundColor:\"white\"});\n",
       "$3Dmol.download(\"pdb:6LU7\", viewer_15888699724689698, {}, function() {\n",
       "\tviewer_15888699724689698.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\n",
       "\tviewer_15888699724689698.render_image();\n",
       "viewer_15888699724689698.render();\n",
       "})\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x7f20a8510d10>"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "view1.render_image()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
